api/websocket/messagebuffer/messagebuffer.go

```diff
@@ -2,38 +2,236 @@ package messagebuffer
+		res = 0
+	case aInt > bInt:
+		res = 1
+	}
+	return
 }
 
 // NewMessageBuffer creates a MessageBuffer
 func NewMessageBuffer() *MessageBuffer {
-	return &MessageBuffer{
-		buffer: make(map[string][]*pb.Relay),
+
+	msgBuffer := &MessageBuffer{
+		buffer:     make(map[string]*rbt.Tree), // *pb.Relay
+		treeCreate: rbt.NewWith(compare),
+		treeExpire: rbt.NewWith(compare),
+	}
+
+	go msgBuffer.BufferMonitor()
+
+	return msgBuffer
+}
+
+// get next sequence id
+func (messageBuffer *MessageBuffer) getNextSeq() int64 {
+
+	messageBuffer.seq.Lock()
+	defer messageBuffer.seq.Unlock()
+	messageBuffer.seq.seqId++
+
+	if messageBuffer.seq.seqId >= maxSeqId {
+		messageBuffer.seq.seqId = 0
 	}
+
+	return messageBuffer.seq.seqId
 }
 
 // AddMessage adds a message to message buffer
 func (messageBuffer *MessageBuffer) AddMessage(clientID []byte, msg *pb.Relay) {
+
+	if msg.MaxHoldingSeconds <= 0 { // no need buffer
+		return
+	}
+
+	now := time.Now().UnixMilli()
+	seqId := messageBuffer.getNextSeq()
+	createId := now*maxSeqId + seqId
+	expireId := (now+(int64(msg.MaxHoldingSeconds)*1000))*maxSeqId + seqId
+
+	mNode := &msgNode{
+		msg:      msg,
+		createId: createId,
+		expireId: expireId,
+	}
+
 	clientIDStr := hex.EncodeToString(clientID)
+
 	messageBuffer.Lock()
 	defer messageBuffer.Unlock()
-	messageBuffer.buffer[clientIDStr] = append(messageBuffer.buffer[clientIDStr], msg)
+
+	_, ok := messageBuffer.buffer[clientIDStr]
+	if !ok {
+		messageBuffer.buffer[clientIDStr] = rbt.NewWith(compare)
+	}
+	messageBuffer.buffer[clientIDStr].Put(mNode.createId, mNode)
+
+	messageBuffer.totalBufSize += mNode.GetSize()
+	// add it to index tree
+	idxNode := &indexNode{clientID: clientIDStr, expireId: expireId, createId: createId}
+	messageBuffer.treeCreate.Put(createId, idxNode)
+	messageBuffer.treeExpire.Put(expireId, idxNode)
+
 }
 
 // PopMessages reads and clears all messages of a client
 func (messageBuffer *MessageBuffer) PopMessages(clientID []byte) []*pb.Relay {
 	clientIDStr := hex.EncodeToString(clientID)
+
 	messageBuffer.Lock()
 	defer messageBuffer.Unlock()
-	messages := messageBuffer.buffer[clientIDStr]
-	messageBuffer.buffer[clientIDStr] = nil
-	return messages
+
+	treeMsg := messageBuffer.buffer[clientIDStr]
+	n := treeMsg.Size()
+
+	relayMessage := make([]*pb.Relay, n, n)
+	i := 0
+	for node := treeMsg.Left(); node != nil; node = treeMsg.Left() { // delete them from index tree
+		mNode := node.Value.(*msgNode)
+		treeMsg.Remove(mNode.createId)
+		messageBuffer.treeCreate.Remove(mNode.createId)
+		messageBuffer.treeExpire.Remove(mNode.expireId)
+		messageBuffer.totalBufSize -= mNode.GetSize()
+
+		relayMessage[i] = mNode.msg
+		i++
+	}
+
+	return relayMessage
+}
+
+// remove expired message
+func (messageBuffer *MessageBuffer) removeExpired() (removedCount int) {
+	now := time.Now().UnixMilli() + 1 // plus 1 milli-second
+	nowId := now * maxSeqId
+
+	removedCount = 0
+	minNode := messageBuffer.treeExpire.Left()
+	for minNode != nil {
```

Comment:
This line is using different style with L143. Better to be consistent in coding style